pkg/services/dashboards/service/zanzana.go

```diff
@@ -79,7 +85,116 @@ func (dr *DashboardServiceImpl) findDashboardsZanzanaCompare(ctx context.Context
+	return dr.findDashboardsZanzanaCheck(ctx, query)
+}
+
+// findDashboardsZanzanaCheck implements "Search, then check" strategy. It first performs search query, then filters out results
+// by checking access to each item.
+func (dr *DashboardServiceImpl) findDashboardsZanzanaCheck(ctx context.Context, query dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
+	ctx, span := tracer.Start(ctx, "dashboards.service.findDashboardsZanzanaCheck")
+	defer span.End()
+
+	result := make([]dashboards.DashboardSearchProjection, 0, query.Limit)
+	var page int64 = 1
+	query.SkipAccessControlFilter = true
+	// Remember initial query limit
+	limit := query.Limit
+	// Set limit to default to prevent pagination issues
+	query.Limit = defaultQueryLimit
+
+	for len(result) < int(limit) {
+		query.Page = page
+		findRes, err := dr.dashboardStore.FindDashboards(ctx, &query)
+		if err != nil {
+			return nil, err
+		}
+
+		remains := limit - int64(len(result))
+		res, err := dr.checkDashboardsBatch(ctx, query, findRes, remains)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, res...)
+		page++
+
+		// Stop when last page reached
+		if len(findRes) < defaultQueryLimit {
+			break
+		}
+	}
+
+	return result, nil
+}
+
+func (dr *DashboardServiceImpl) checkDashboardsBatch(ctx context.Context, query dashboards.FindPersistedDashboardsQuery, searchRes []dashboards.DashboardSearchProjection, remains int64) ([]dashboards.DashboardSearchProjection, error) {
+	ctx, span := tracer.Start(ctx, "dashboards.service.checkDashboardsBatch")
+	defer span.End()
+
+	if len(searchRes) == 0 {
+		return nil, nil
+	}
+
+	batchReqItems := make([]*authzextv1.BatchCheckItem, 0, len(searchRes))
+
+	for _, d := range searchRes {
+		// FIXME: support different access levels
+		kind := zanzana.KindDashboards
+		action := dashboards.ActionDashboardsRead
+		if d.IsFolder {
+			kind = zanzana.KindFolders
+			action = dashboards.ActionFoldersRead
+		}
+
+		checkReq, ok := zanzana.TranslateToCheckRequest("", action, kind, d.FolderUID, d.UID)
+		if !ok {
+			continue
+		}
+
+		batchReqItems = append(batchReqItems, &authzextv1.BatchCheckItem{
+			Verb:        checkReq.Verb,
+			Group:       checkReq.Group,
+			Resource:    checkReq.Resource,
+			Name:        checkReq.Name,
+			Folder:      checkReq.Folder,
+			Subresource: checkReq.Subresource,
+		})
+	}
+
+	batchReq := authzextv1.BatchCheckRequest{
+		Namespace: query.SignedInUser.GetNamespace(),
+		Subject:   query.SignedInUser.GetUID(),
+		Items:     batchReqItems,
+	}
+
+	res, err := dr.zclient.BatchCheck(ctx, &batchReq)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]dashboards.DashboardSearchProjection, 0)
+	for _, d := range searchRes {
+		if len(result) >= int(remains) {
+			break
+		}
+
+		kind := zanzana.KindDashboards
+		if d.IsFolder {
+			kind = zanzana.KindFolders
+		}
+		resourceInfo, ok := zanzana.TranslateResource(kind)
+		if !ok {
+			continue
+		}
```

Comment:
This is a correct way now, but since we know that dashboard/folder UIDs are unique, maybe we can remove it. It looks a bit ugly.